### ISSUE TITLE
Built-in `blog` plugin give `AssertionError` with custom index pages

### ISSUE THREAD

**Issue by balansczerni** [squidfunk/mkdocs-material#5896]
### Context

I'm trying to create a custom index page for blog categories.

### Bug description

After creating a custom index page (*"custom-category-page.md"*) for any of the blog categories (category: *"custom-category-page"*), I get the error:

```
INFO    -  Building documentation...
INFO    -  Cleaning site directory
Traceback (most recent call last):
  File "/Users/name/.pyenv/versions/3.11.4/bin/mkdocs", line 8, in <module>
    sys.exit(cli())
             ^^^^^
  File "/Users/name/.pyenv/versions/3.11.4/lib/python3.11/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/name/.pyenv/versions/3.11.4/lib/python3.11/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/Users/name/.pyenv/versions/3.11.4/lib/python3.11/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/name/.pyenv/versions/3.11.4/lib/python3.11/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/name/.pyenv/versions/3.11.4/lib/python3.11/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/name/.pyenv/versions/3.11.4/lib/python3.11/site-packages/mkdocs/__main__.py", line 270, in serve_command
    serve.serve(**kwargs)
  File "/Users/name/.pyenv/versions/3.11.4/lib/python3.11/site-packages/mkdocs/commands/serve.py", line 86, in serve
    builder(config)
  File "/Users/name/.pyenv/versions/3.11.4/lib/python3.11/site-packages/mkdocs/commands/serve.py", line 67, in builder
    build(config, live_server=None if is_clean else server, dirty=is_dirty)
  File "/Users/name/.pyenv/versions/3.11.4/lib/python3.11/site-packages/mkdocs/commands/build.py", line 311, in build
    nav = config.plugins.on_nav(nav, config=config, files=files)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/name/.pyenv/versions/3.11.4/lib/python3.11/site-packages/mkdocs/plugins.py", line 536, in on_nav
    return self.run_event('nav', nav, config=config, files=files)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/name/.pyenv/versions/3.11.4/lib/python3.11/site-packages/mkdocs/plugins.py", line 507, in run_event
    result = method(item, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/name/.pyenv/versions/3.11.4/lib/python3.11/site-packages/material/plugins/blog/plugin.py", line 174, in on_nav
    views = [*self._generate_categories(config, files)]
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/name/.pyenv/versions/3.11.4/lib/python3.11/site-packages/material/plugins/blog/plugin.py", line 597, in _generate_categories
    assert isinstance(file.page, Category)
AssertionError
```

Related code: (597-599)
```
                # Assign post to category and vice versa
                assert isinstance(file.page, Category)
                file.page.posts.append(post)
                post.categories.append(file.page)
```

---

The error appears whether I modify `categories_url_format:`. I got the same results with: 

`categories_url_format: {slug}` (in my case)

```
.
├─ docs/
│  └─ blog/
│     └─ custom-category-page.md 
```

as  with default `categories_url_format: category/{slug}`

```
.
├─ docs/
│  └─ blog/
│     ├─ category/
│     │  └─ custom-category-page.md 
```

---

According to [the current documentation](https://squidfunk.github.io/mkdocs-material/setup/setting-up-a-blog/#custom-index-pages), this should allow me to add content to category pages:

> If you want to add custom content to automatically generated [archive](https://squidfunk.github.io/mkdocs-material/setup/setting-up-a-blog/#archive) and [category](https://squidfunk.github.io/mkdocs-material/setup/setting-up-a-blog/#categories) indexes, e.g. to add a category description prior to the list of posts, you can manually create the category page in the same location where the [built-in blog plugin](https://squidfunk.github.io/mkdocs-material/setup/setting-up-a-blog/#built-in-blog-plugin) would create it.

### Related links

- [Custom index pages](https://squidfunk.github.io/mkdocs-material/setup/setting-up-a-blog/#custom-index-pages)
- [similar issue](https://github.com/squidfunk/mkdocs-material/issues/5880) #5880 
- [similar issue](https://github.com/squidfunk/mkdocs-material/issues/5879) #5879

### Reproduction

* `. venv/bin/activate`
* `pip install --upgrade --force-reinstall mkdocs-material`
* `pip install git+https://${GH_TOKEN}@github.com/squidfunk/mkdocs-material-insiders.git`
* `mkdocs new .`
* `mkdocs build`

Minimal reproduction: [insiders-custom-index-page-fail-raw.zip](https://github.com/squidfunk/mkdocs-material/files/12440597/insiders-custom-index-page-fail-raw.zip)

additional version with modified `categories_url_format` settings which **do not fix** the error:

[insiders-custom-index-page-fail.zip](https://github.com/squidfunk/mkdocs-material/files/12440522/insiders-custom-index-page-fail.zip)

### Steps to reproduce

The bug affects the **insiders** version related features ([blog customization](https://squidfunk.github.io/mkdocs-material/setup/setting-up-a-blog/#customization)).

1. Create blog with build-in blugin
2. Create post with category name (ex. "custom")
3. Follow [custom index page doc.](https://squidfunk.github.io/mkdocs-material/setup/setting-up-a-blog/#custom-index-pages):
   1. Take the URL generated by the built-in blog plugin (*/blog/category/custom/*)
   2. Create page with the same name (*custom.md*) in category directory as follows.
   3. Try build/serve again.

```
.
├─ docs/
│  └─ blog/
│     ├─ category/
│     │  └─ custom.md 
│     ├─ posts/
│     └─ index.md
└─ mkdocs.yml
```

### Before submitting

- [X] I have read and followed the [bug reporting guidelines](https://squidfunk.github.io/mkdocs-material/contributing/reporting-a-bug/).
- [X] I have attached links to [the documentation](https://squidfunk.github.io/mkdocs-material), and possibly related [issues](https://github.com/squidfunk/mkdocs-material/issues) and [discussions](https://github.com/squidfunk/mkdocs-material/discussions).
- [X] I assure that I have [removed all customizations](https://squidfunk.github.io/mkdocs-material/contributing/reporting-a-bug/#remove-customizations) before submitting this bug report.
- [X] I have attached a __.zip file__ with a [minimal reproduction](https://squidfunk.github.io/mkdocs-material/guides/creating-a-reproduction/).

**Comment by squidfunk** [comment]
Thanks for reporting! This is definitely a regression of the blog plugin rewrite to make it future-proof. I'll look into it asap. In the meantime, the issue can be mitigated by downgrading Insiders to [9.1.21-insiders-4.39.0](https://github.com/squidfunk/mkdocs-material-insiders/releases/tag/9.1.21-insiders-4.39.0), if you need it fixed now.

**Comment by squidfunk** [comment]
Thanks again for the excellent reproduction and digging into the topic. Fixed in [`9667f55e9`](https://github.com/squidfunk/mkdocs-material-insiders/commit/9667f55e97bedcc9c5ee9beed36ea61c3657f886).

**Comment by squidfunk** [comment]
Released as part of [9.2.4+insiders-4.40.0](https://squidfunk.github.io/mkdocs-material/insiders/changelog/#4.40.0).

**Comment by balansczerni** [comment]
Thanks for appreciating (this is my first bug report ever and I tried my best).

---

**Unfortunately I have to inform you that this fix is not perfect and causes a different error. 🫡**

(I upgrade with `pip install --upgrade git+https://${GH_TOKEN}@github.com/squidfunk/mkdocs-material-insiders.git` and made sure that *blog/plugin.py* have [9667f55e9](https://github.com/squidfunk/mkdocs-material-insiders/commit/9667f55e97bedcc9c5ee9beed36ea61c3657f886) changes)

Everything works correctly *only if* a given category has *only one* post. 

Bug1: When we add another post to the same category, the "None" object will appear in the navigation (which links to the category page again). Each subsequent post adds such "None" element.

Bug2: Only one entry is displayed on the custom category page. Subsequent entries are assigned to the "None" category, although their URL is correct.

**Reproduction:**
[none-items.zip](https://github.com/squidfunk/mkdocs-material/files/12445301/none-items.zip)

<img width="654" alt="Zrzut ekranu 2023-08-26 o 13 51 47" src="https://github.com/squidfunk/mkdocs-material/assets/123209821/2c1f057a-a700-4e2e-ad8d-732736e317e1">

---

`pip list`

```
mkdocs                     1.5.2
mkdocs-material            9.2.3+insiders.4.39.3
mkdocs-material-extensions 1.1.1
```

*MkDocs.yml:*

```
nav:
  - Home: index.md
  - Blog:
    - blog/index.md
    - blog/category/custom-category-page.md
```


```
.
├─ docs/
│  └─ blog/
│     ├─ category/
│     │  └─ custom-category-page.md 
│     ├─ posts/
|     |  ├─ example-2.md
|     |  └─ example.md 
│     └─ index.md
└─ mkdocs.yml
```

---

Should I report this as a new bug or can it stay here?

**Comment by squidfunk** [comment]
Thanks for testing. The behavior you're observing now are definitely different bugs, so please create a new issue (I think both observations are related), so other users can discover them more easily. We're happy to fix them.

**Comment by squidfunk** [comment]
Additionally, `pip list` shows `9.2.3+insiders.4.39.3` – make sure to upgrade to `9.2.4+insiders.4.40.0` before.

**Comment by squidfunk** [comment]
Wait, I think I already got it. Potential fix in [`9d8e1c5`](https://github.com/squidfunk/mkdocs-material-insiders/commit/9d8e1c5ea3f1eb54936f70a0ef3e7d6b1f62d949). Note that you should not link the category pages in `mkdocs.yml` in `nav` – the blog plugin will link them for you.

![screenshot-localhost-8000-blog-category-custom-category-page-1693054189713](https://github.com/squidfunk/mkdocs-material/assets/932156/2ae5541f-774e-4b41-a39d-59b208766596)

**Comment by squidfunk** [comment]
Released as part of [9.2.5+insiders-4.40.1](https://squidfunk.github.io/mkdocs-material/insiders/changelog/#4.40.1).
